### PR TITLE
Upgrade `openapi-typescript` to 7.0.2 + swap to `openapi-fetch`

### PR DESCRIPTION
**Pull request by davelopez**
Upgrade to the latest version of `openapi-typescript`. This introduces some [breaking changes](https://github.com/openapi-ts/openapi-typescript/releases/tag/openapi-typescript%407.0.0) in particular:

> Most optional objects are now always present in types, just typed as :never. This includes keys of the Components Object as well as HTTP methods.

Maybe this is a good opportunity to continue with @dannon's work started in https://github.com/dannon/galaxy/commits/openapi-typescript-fetch-upgrade/ and swap to [openapi-fetch](https://github.com/openapi-ts/openapi-typescript/tree/main/packages/openapi-fetch)

## Update

See discussion details at https://github.com/galaxyproject/galaxy/discussions/18534

TLDR; There are some more incompatibilities between the latest `openapi-typescript` generated schemas and the current `openapi-typescript-fetcher` so, if we want to upgrade we should replace the current fetcher with the most up-to-date `openapi-fetch` variant.

## TODO
- [x] Replace `fetcher` with `client`
- [x] Improve a bunch of things in the API schema
  - [x] https://github.com/galaxyproject/galaxy/pull/18563
  - [x] https://github.com/galaxyproject/galaxy/pull/18586
  - [x] https://github.com/galaxyproject/galaxy/pull/18599
  - [x] https://github.com/galaxyproject/galaxy/pull/18602
  - [x] https://github.com/galaxyproject/galaxy/pull/18618
  - [x] https://github.com/galaxyproject/galaxy/pull/18645
  - [x] https://github.com/galaxyproject/galaxy/pull/18668
- [x] Fix all unit tests and mocks (seems to work in recent node versions)
- [x] Update https://github.com/galaxyproject/galaxy/blob/dev/client/docs/querying-the-api.md

---

# New way of querying the API

For more information see the [updated client docs](https://github.com/galaxyproject/galaxy/blob/9c0708d8cfb970b8313b25dd0b1f83cc445b4536/client/docs/querying-the-api.md).

```ts
import { ref, onMounted } from "vue";

import { GalaxyApi, type HDADetailed } from "@/api";
import { errorMessageAsString } from "@/utils/simple-error";

interface Props {
    datasetId: string;
}

const props = defineProps<Props>();

const datasetDetails = ref<HDADetailed>();
const errorMessage = ref<string>();

async function loadDatasetDetails() {
    // Your IDE will provide you with autocompletion for the route and all the parameters
    const { data, error } = await GalaxyApi().GET("/api/datasets/{dataset_id}", {
        params: {
            path: {
                dataset_id: props.datasetId,
            },
            query: { view: "detailed" },
        },
    });

    if (error) {
        // Handle error here. For example, you can display a message to the user.
        errorMessage.value = errorMessageAsString(error);
        // Make sure to return here, otherwise `data` will be undefined
        return;
    }

    // Use `data` here. We are casting it to HDADetailed to help the type inference because
    // we requested the "detailed" view and this endpoint returns different types depending
    // on the view. In general, the correct type will be inferred automatically using the
    // API schema so you don't need to (and shouldn't) cast it.
    datasetDetails.value = data as HDADetailed;
}

onMounted(() => {
    loadDatasetDetails();
});
```

# New way of mocking the API responses in the client unit test

Along with the openapi-fetch replacement, we introduced a new way of mocking API calls in the client using [Mock Service Worker](https://mswjs.io/).

For more information see the [updated docs](https://github.com/galaxyproject/galaxy/blob/9c0708d8cfb970b8313b25dd0b1f83cc445b4536/client/docs/unit-testing/writing-tests.md).

```ts
import { useServerMock } from "@/api/client/__mocks__";

const { server, http } = useServerMock();

describe("MyComponent", () => {
    it("should do something with the history", async () => {
        // Mock the response of the API call
        server.use(
            http.get("/api/histories/{history_id}", ({ params, query, response }) => {
                // You can use logic to return different responses based on the request
                if (query.get("view") === "detailed") {
                    return response(200).json(TEST_HISTORY_DETAILED);
                }

                // Or simulate an error
                if (params.history_id === "must-fail") {
                    return response("5XX").json(EXPECTED_500_ERROR, { status: 500 });
                }

                return response(200).json(TEST_HISTORY_SUMMARY);
            })
        );

        // Your test code here
    });
});
```

## How to test the changes?
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

**Comment by mvdbeek** [comment]
I think this looks pretty good, what's a clear improvement is that default values and constants don't automatically make attributes optional:
```diff
              * Ext
              * @default auto
              */
-            ext?: string;
+            ext: string;
```

**Comment by davelopez** [comment]
Should I give it a try to swap to `openapi-fetch` directly here then? Or try to fix all the "never" issues in the current `openapi-typescript-fetch`? :thinking:

**Comment by mvdbeek** [comment]
I don't know what's better, I assume all these things need to be fixed whether we use openapi-fetch or not ?

**Comment by davelopez** [comment]
> I assume all these things need to be fixed whether we use openapi-fetch or not ?

Yes, but if we plan to move to `openapi-fetch` in the future, we will have to replace most if not all fetcher calls anyway so it would be fixing something that we are going to replace later.

I think I will give it a try to replace it, and if it is too disruptive I will fall back and just fix the typing errors.

**Comment by dannon** [comment]
@davelopez please do just go ahead with that and swap over; I think it's the right move and I'd be more than happy for you to take over!

File: client/src/api/datasetCollections.ts [hunk 1]
```diff
@@ -19,15 +24,19 @@ export async function fetchCollectionDetails(params: { id: string }): Promise<HD
...
+    const { data, error } = await client.GET("/api/dataset_collections/{id}", {
+        params: {
+            path: params,
+            query: { view: "collection" },
+        },
+    });
+
+    if (error) {
+        rethrowSimple(error);
+    }
```
Comment (mvdbeek) [inline client/src/api/datasetCollections.ts:35]: Could we instead register a middleware like we did with the old fetcher ?

Comment (davelopez) [reply client/src/api/datasetCollections.ts:35]: Yes, I have considered it: https://github.com/galaxyproject/galaxy/pull/18532/files#diff-a29bb8e3afc7a891c203b6960df3ff89149f8ef3e4e80658d8a9e35a53d5620aR24

I'll work on that next. I wanted to ask about it in the meeting but couldn't make it. The thing is, I was still unsure because one of the differences between `openapi-fetch` and `openapi-typescript-fetch` is:

> openapi-typescript-fetch throws exceptions for non-OK responses (requiring you to wrap things in try/catch) rather than follow the fetch API spec of not throwing.

Something like this:
### Do not throw
```ts
const { data, error } = await client.GET("/api/libraries");
// here data can be undefined if error is defined. If you try to use it it will fail type validation
// so you are forced to handle the error at "type checking time"
if (error) {
    // handle error, return or throw
}
// data is guaranteed to be defined here
```

So if we add the middleware to rethrow like we were doing, we now probably need to wrap the call in try/catch and, at the same time, we still have to deal with `data` being possibly undefined when `error` is defined like this:

### Throw in middleware
```ts
try{
    const { data, error } = await client.GET("/api/libraries");
    // here data can still be undefined if error is defined
    if (error) {
        // handle error, return or throw
    }
    // data is guaranteed to be defined here
}catch(e){
    // handle error (again?)
}
```

Maybe there is a way to tell the middleware to assume `data` will always be defined as the exception would have been thrown otherwise?

I'm right now a bit on the fence but slightly leaning towards the first option, because it doesn't require you to remember to wrap the call in a try/catch, and on the other hand, it forces you to deal with the error early, otherwise, typescript will complain. Of course, at the cost of calling rethrowSimple or similar on the spot instead of in a single place...

Comment (mvdbeek) [reply client/src/api/datasetCollections.ts:35]: If we're committed to handling the optional type and the error in the calling stack that would be a solid improvement for sure 👍. On a quick glance I just saw 

```
const { data, error } = await client.GET("/api/libraries");
// here data can be undefined if error is defined. If you try to use it it will fail type validation
// so you are forced to handle the error at "type checking time"
if (error) {
    rethrowSimple(error)
}
```
in the PR which is why I thought you'd opted for handling errors like we used to, in which case the middleware is less verbose.

Comment (davelopez) [reply client/src/api/datasetCollections.ts:35]: Yes, those are some of the wrapper functions from "@api/*" that are reused in multiple places. I choose to keep those that were used in more than one place. I'm forced to rethrow there because we don't have the context to handle the error properly and the caller code should be already expecting exceptions.

There are also some situations in which error handling was not considered at all in the calling code or already had a try/catch, in those cases I'm also simply rethrowing to keep the changes minimal. In those cases where I clearly see how to handle the error I'm removing the try/catch and adapting accordingly.

**Comment by davelopez** [comment]
This is almost ready, all fetcher call have been replaced and the tests updated, but there are still some things that I can't get to work:

- The client unit tests in the CI don't run. It seems is missing `undici` dependency, but if I install it, then it breaks the unit tests locally for me... the whole configuration around `jest`, `jsdom` and `msw` might not be optimal here... I'm a bit lost with it.
- There are a bunch of unit tests failing, apparently for async issues, some await etc must be missing somewhere but it is hard to sport, I will keep investigating.

**Comment by davelopez** [comment]
I think this is finally ready for review. Although not ready to be merged so I'll keep it in draft for now. But it would be good to have others reviews and address potential changes soon. If you prefer a different naming convention, or other improvements, etc, please suggest :smiley: 

The client unit tests are passing for me locally using node v20.16.0 I guess they are failing in the CI because some of them are not handling timers and async promise rejections properly. 

I'll take a look at them and see if I can fix those hanging promises. In general, we should ensure our jest tests do not have so many ignored warnings and errors (even if the test passes OK) because it makes it hard to debug these issues :disappointed:

File: client/docs/querying-the-api.md [hunk 2]
```diff
@@ -24,19 +24,32 @@ If there is no Composable for the API endpoint you are using, try using a (Pinia
...
+        path: {
+            dataset_id: datasetId,
+        },
+        query: { view: "detailed" },
+    },
+});
+
+if (error) {
+    // Handle error here and return otherwise `data` will be undefined
+    return;
```
Comment (mvdbeek) [inline client/docs/querying-the-api.md]: I get that's not the intention, but if someone takes this example literally that is exactly what we should avoid 😆.
Maybe we can have a real example here ?

Comment (davelopez) [reply client/docs/querying-the-api.md]: You are right, lazy doc writing... is worse than no doc at all sometimes :sweat_smile:

Comment (davelopez) [reply client/docs/querying-the-api.md]: Is something like https://github.com/galaxyproject/galaxy/pull/18532/commits/1ffd38242bb4926d2a0456dd05cdedd078520849 slightly better?

File: client/src/components/DatasetInformation/DatasetDetails.vue [hunk 3]
```diff
@@ -41,42 +41,43 @@ const datasetLoadingError = ref<string | null>(null);
...
-        const error = e as AxiosError<{ err_msg?: string }>;
+    const { data, error } = await GalaxyApi().GET("/api/jobs/{job_id}", {
+        params: {
+            path: { job_id: dataset.value?.creating_job! },
+            query: { full: true },
+        },
+    });
+
+    if (error) {
+        jobLoadingError.value = errorMessageAsString(error);
```
Comment (mvdbeek) [inline client/src/components/DatasetInformation/DatasetDetails.vue:63]: This seems like a better example to me @davelopez. Toasts are simple, but they often lack context, especially if there's more than one possible error on a page.

File: client/src/components/JobInformation/JobInformation.vue [hunk 4]
```diff
@@ -59,7 +59,11 @@ function filterMetadata(jobMessages) {
 
 async function fetchInvocation(jobId) {
     if (jobId) {
-        const invocation = await invocationForJob({ jobId: jobId });
+        const { data: invocation } = await GalaxyApi().GET("/api/invocations", {
+            params: {
+                query: { job_id: jobId },
+            },
+        });
```
Comment (mvdbeek) [inline client/src/components/JobInformation/JobInformation.vue:67]: We're missing some kind of error handling here, no ?

File: client/src/components/ObjectStore/Instances/EditInstance.vue [hunk 5]
```diff
@@ -36,8 +37,21 @@ const loadingMessage = "Loading storage location template and instance informati
...
+        const { data: objectStore, error } = await GalaxyApi().PUT(
+            "/api/object_store_instances/{user_object_store_id}",
+            {
+                params: { path: { user_object_store_id } },
+                body: payload,
+            }
+        );
+
+        if (error) {
+            rethrowSimple(error);
```
Comment (mvdbeek) [inline client/src/components/ObjectStore/Instances/EditInstance.vue:52]: This seems to be only handled at https://github.com/galaxyproject/galaxy/blob/dev/client/src/entry/analysis/router.js#L739-L743, if this component is ever mounted as a subcomponent there would be no error handling at all, I think ?
All of this to say maybe this needs some error handling inside of the component ?

Comment (mvdbeek) [reply client/src/components/ObjectStore/Instances/EditInstance.vue:52]: Which doesn't need to happen now, it isn't worse than before.

File: lib/tool_shed/webapp/frontend/src/components/pages/CitableRepositoryPage.vue [hunk 6]
```diff
@@ -18,7 +16,14 @@ const error = ref<string | null>(null)
 
 async function update() {
     error.value = null
-    const { data } = await indexFetcher({ owner: props.username, name: props.repositoryName })
+    const { data } = await client.GET("/api/repositories", {
```
Comment (mvdbeek) [inline lib/tool_shed/webapp/frontend/src/components/pages/CitableRepositoryPage.vue]: We've lost all error handling here, no? The old middleware would have thrown on error.

File: lib/tool_shed/webapp/frontend/src/components/pages/ManageApiKey.vue [hunk 7]
```diff
@@ -26,15 +22,28 @@ async function copyKey() {
...
-    apiKeyFetcher(params)
+    client
+        .GET("/api/users/{encoded_user_id}/api_key", {
+            params: {
+                path: params,
+            },
+        })
         .then(({ data }) => {
+            if (!data) {
+                throw Error("Error fetching API key")
```
Comment (mvdbeek) [inline lib/tool_shed/webapp/frontend/src/components/pages/ManageApiKey.vue]: Shouldn't we rethrowSimple the error here ?

File: client/src/components/DatasetInformation/DatasetDetails.vue (hunk 3 above)
Comment (davelopez) [reply client/src/components/DatasetInformation/DatasetDetails.vue:63]: Thanks for the suggestion, I'm adding a more complete example with some extra information about advantages/disadvantages.

File: client/src/components/JobInformation/JobInformation.vue (hunk 4 above)
Comment (davelopez) [reply client/src/components/JobInformation/JobInformation.vue:67]: Well spotted I'll add at least a rethrowSimple here :+1:

We can follow up with more improved error handling in general looking for all the rethrowSimple calls :sweat_smile:

File: client/src/components/ObjectStore/Instances/EditInstance.vue (hunk 5 above)
Comment (davelopez) [reply client/src/components/ObjectStore/Instances/EditInstance.vue:52]: I agree, in general, we need better error handling in most components. I didn't want to add many changes to the already monstrous PR, but we should definitely be able to search for those specific `rethrowSimple` calls and maybe transform them to better error handling in the components as a follow-up. At least it is easier to find those spots now :smiley:

File: lib/tool_shed/webapp/frontend/src/components/pages/CitableRepositoryPage.vue (hunk 6 above)
Comment (davelopez) [reply lib/tool_shed/webapp/frontend/src/components/pages/CitableRepositoryPage.vue]: Good catch again :)
Fixing it...

File: lib/tool_shed/webapp/frontend/src/components/pages/ManageApiKey.vue (hunk 7 above)
Comment (davelopez) [reply lib/tool_shed/webapp/frontend/src/components/pages/ManageApiKey.vue]: I'm afraid we don't have `rethrowSimple` in the context of the Tool Shed client... I was hoping this was handled by the `notifyOnCatch` handler below. I can still refactor it for async/await syntax for better readability.

Comment (mvdbeek) [reply lib/tool_shed/webapp/frontend/src/components/pages/ManageApiKey.vue]: It (probably) is, I was more thinking that we might be missing out on valuable server-side error messages if we cast all errors to this generic string. Can't we just import rethrowSimple ?

Comment (davelopez) [reply lib/tool_shed/webapp/frontend/src/components/pages/ManageApiKey.vue]: Ahh, I see your point now! Thanks for clarifying!

Comment (davelopez) [reply lib/tool_shed/webapp/frontend/src/components/pages/ManageApiKey.vue]: AFAIK we can't import rethrowSimple because the Tool Shed 2.0 client code is now completely separated from Galaxy, but I can use the proper `error_msg` content for the error reason.

File: client/src/components/DatasetInformation/DatasetDetails.vue (hunk 3 above)
Comment (davelopez) [reply client/src/components/DatasetInformation/DatasetDetails.vue:63]: Done! https://github.com/galaxyproject/galaxy/pull/18532/commits/b8be283f6ee3a6eb08186f83acd6ffc4fcae7458

File: lib/tool_shed/webapp/frontend/src/components/pages/CitableRepositoryPage.vue (hunk 6 above)
Comment (davelopez) [reply lib/tool_shed/webapp/frontend/src/components/pages/CitableRepositoryPage.vue]: Done https://github.com/galaxyproject/galaxy/pull/18532/commits/094e28c3ed529820d2beae1ee62a931cccb209e6

File: lib/tool_shed/webapp/frontend/src/components/pages/ManageApiKey.vue (hunk 7 above)
Comment (davelopez) [reply lib/tool_shed/webapp/frontend/src/components/pages/ManageApiKey.vue]: Done! https://github.com/galaxyproject/galaxy/pull/18532/commits/9c0708d8cfb970b8313b25dd0b1f83cc445b4536

File: client/src/components/JobInformation/JobInformation.vue (hunk 4 above)
Comment (davelopez) [reply client/src/components/JobInformation/JobInformation.vue:67]: Done! https://github.com/galaxyproject/galaxy/pull/18532/commits/37dcdf4a58cae25820bdb038e10dd349f8f2cf92

**Comment by davelopez** [comment]
Finally! all client unit tests are green even using node v18 :tada: 

The integration test failures I think are the same old ones that fail in any other PR

File: lib/tool_shed/webapp/frontend/src/components/pages/ManageApiKey.vue (hunk 7 above)
Comment (mvdbeek) [reply lib/tool_shed/webapp/frontend/src/components/pages/ManageApiKey.vue]: But then we're missing out on proxy errors like gateway timeout or things like that if we assume all errors are in `err_msg`. If there's no way to share code with the Galaxy client I would copy-paste rethrowSimple (or install the galaxy client from npm ?)

File: client/src/components/FileSources/Instances/EditInstance.vue [hunk 8]
```diff
@@ -36,8 +37,17 @@ const loadingMessage = "Loading file source template and instance information";
 async function onSubmit(formData: any) {
     if (template.value) {
         const payload = editFormDataToPayload(template.value, formData);
-        const args = { user_file_source_id: String(instance?.value?.uuid) };
-        const { data: fileSource } = await update({ ...args, ...payload });
+        const user_file_source_id = instance?.value?.uuid!;
```
Comment (mvdbeek) [inline client/src/components/FileSources/Instances/EditInstance.vue]: Does it compile if you do `if (template.value && instance.value)` and then remove the non-null operator ?
It seems a little clearer that way that onSubmit is only called if both of these are defined (though a comment wouldn't hurt).

File: client/src/components/History/Archiving/HistoryArchive.vue [hunk 9]
```diff
@@ -50,16 +48,24 @@ async function updateSearchQuery(query: string) {
...
+            currentPage: currentPage.value,
+            pageSize: perPage.value,
+            sortBy: sortBy.value,
+            sortDesc: sortDesc.value,
+        });
+        totalRows.value = result.totalMatches;
+        archivedHistories.value = result.histories;
+    } catch (error) {
+        toast.error(
+            localize(`Failed to load archived histories with reason: ${errorMessageAsString(error)}`),
```
Comment (mvdbeek) [inline client/src/components/History/Archiving/HistoryArchive.vue:63]: Not for this PR, but maybe we should bake that into `errorMessageAsString`

File: lib/tool_shed/webapp/frontend/src/components/pages/ManageApiKey.vue (hunk 7 above)
Comment (davelopez) [reply lib/tool_shed/webapp/frontend/src/components/pages/ManageApiKey.vue]: Hmm... that is true, however, it seems I kept a middleware for error handling when I was experimenting with the Tool Shed client, will that cover that case?

https://github.com/galaxyproject/galaxy/pull/18532/files#diff-266bc08bc3bd8a93c2932dda21400afe346a811fb19d093b250f5925796a6416

Certainly, the Toolshed code was the first I migrated while learning, so it can benefit from some more refactoring having the new fetcher design in mind... should we give it a try in this PR or better in a follow-up?

File: client/src/components/admin/Notifications/NotificationForm.vue [hunk 10]
```diff
@@ -99,9 +86,37 @@ async function loadData<T>(
     try {
         const tmp = await getData();
         target.value = tmp.map(formatter);
-    } catch (error: any) {
+    } catch (error) {
+        Toast.error(errorMessageAsString(error));
```
Comment (mvdbeek) [inline client/src/components/admin/Notifications/NotificationForm.vue]: I think you can remove this since you're not throwing in any of the functions passed in as getData ?

File: client/src/stores/configurationStore.ts [hunk 11]
```diff
@@ -17,11 +15,14 @@ export const useConfigStore = defineStore("configurationStore", () => {
         if (!isLoaded.value && !isLoading.value) {
             isLoading.value = true;
             try {
-                const { data } = await fetchConfiguration({});
+                const { data, error } = await GalaxyApi().GET("/api/configuration");
+
+                if (error) {
+                    console.error("Error loading Galaxy configuration", error);
```
Comment (mvdbeek) [inline client/src/stores/configurationStore.ts:21]: 😢 In a follow up I would say we should add another property to the store that holds the error and makes the data possibly undefined.

File: client/src/components/History/Archiving/HistoryArchive.vue (hunk 9 above)
Comment (davelopez) [reply client/src/components/History/Archiving/HistoryArchive.vue:63]: You mean the `localize` for all error messages right? :+1:

File: client/src/components/FileSources/Instances/EditInstance.vue (hunk 8 above)
Comment (davelopez) [reply client/src/components/FileSources/Instances/EditInstance.vue]: Yep, that was an oversight. Thanks for the catch!

File: lib/tool_shed/webapp/frontend/src/components/pages/ManageApiKey.vue (hunk 7 above)
Comment (mvdbeek) [reply lib/tool_shed/webapp/frontend/src/components/pages/ManageApiKey.vue]: The middleware approach in the TS seems fine, but we should do what Galaxy does with either rethrowSimple / errorMessageAsString. I don't think the backend is decomposed so much so that we can rely on more uniform error handling.

File: client/src/components/History/Archiving/HistoryArchive.vue (hunk 9 above)
Comment (mvdbeek) [reply client/src/components/History/Archiving/HistoryArchive.vue:63]: yes

File: lib/tool_shed/webapp/frontend/src/components/RegisterPage.vue [hunk 12]
```diff
@@ -12,19 +12,23 @@ const confirm = ref("")
...
+        const { data } = await client.POST("/api_internal/register", {
+            body: {
+                email: email.value,
+                password: password.value,
+                username: username.value,
+                bear_field: "",
+            },
         })
+        if (!data) {
+            throw new Error("No data returned")
```
Comment (mvdbeek) [inline lib/tool_shed/webapp/frontend/src/components/RegisterPage.vue]: That's similar to https://github.com/galaxyproject/galaxy/pull/18532#discussion_r1713624420

File: lib/tool_shed/webapp/frontend/src/components/RepositoryActions.vue [hunk 13]
```diff
@@ -1,10 +1,12 @@
...
 import { notify, notifyOnCatch } from "@/util"
-const resetFetcher = fetcher.path("/api/repositories/{encoded_repository_id}/reset_metadata").method("post").create()
 
 async function resetMetadata() {
-    resetFetcher({ encoded_repository_id: props.repositoryId })
+    client
+        .POST("/api/repositories/{encoded_repository_id}/reset_metadata", {
+            params: { path: { encoded_repository_id: props.repositoryId } },
+        })
         .catch(notifyOnCatch)
```
Comment (mvdbeek) [inline lib/tool_shed/webapp/frontend/src/components/RepositoryActions.vue:10]: This only works if we keep the middleware approach (which is fine, not saying this needs to change)

File: lib/tool_shed/webapp/frontend/src/components/pages/CitableRepositoryPage.vue [hunk 14]
```diff
@@ -18,8 +16,17 @@ const error = ref<string | null>(null)
...
+    const { data, error: requestError } = await client.GET("/api/repositories", {
+        params: {
+            query: {
+                owner: props.username,
+                name: props.repositoryName,
+            },
+        },
+    })
+    if (requestError) {
+        error.value = requestError.err_msg
```
Comment (mvdbeek) [inline lib/tool_shed/webapp/frontend/src/components/pages/CitableRepositoryPage.vue]: Not sure it's always a JSON response with err_msg

**Review by mvdbeek — APPROVED** [review]
The Galaxy portion looks great, I think for the TS side we need to not assume data.err_msg hold a proper error message.

File: client/src/components/History/Archiving/HistoryArchive.vue (hunk 9 above)
Comment (mvdbeek) [reply client/src/components/History/Archiving/HistoryArchive.vue:63]: I've opened https://github.com/galaxyproject/galaxy/issues/18688 with that

File: lib/tool_shed/webapp/frontend/src/components/pages/CitableRepositoryPage.vue (hunk 14 above)
Comment (davelopez) [reply lib/tool_shed/webapp/frontend/src/components/pages/CitableRepositoryPage.vue]: If we trust the schema it should :sweat_smile: 
But now that I remember I added the middleware probably none of this works as expected because all requests in the Tool Shed need to be wrapped in a try/catch, I think :thinking:

File: lib/tool_shed/webapp/frontend/src/components/RegisterPage.vue (hunk 12 above)
Comment (davelopez) [reply lib/tool_shed/webapp/frontend/src/components/RegisterPage.vue]: I think this line will never be reached because of the middleware. It could be just replaced with an empty return or a check for data is defined before using it.

File: lib/tool_shed/webapp/frontend/src/components/pages/CitableRepositoryPage.vue (hunk 14 above)
Comment (mvdbeek) [reply lib/tool_shed/webapp/frontend/src/components/pages/CitableRepositoryPage.vue]: The schema is for the app, but the proxy can answer with gateway timeouts, rate limits etc

**Comment by davelopez** [comment]
Thanks for the detailed review Marius!
Now that I looked again, the Toolshed migration (4 weeks is too much for my fish memory :sweat_smile:) needs a second round... If we keep the middleware approach I need to wrap every API call with a try/catch, ignore the error, and check for data to be defined...

**Comment by mvdbeek** [comment]
> ignore the error, and check for data to be defined...

can't that be done in the middleware ?

**Comment by davelopez** [comment]
> can't that be done in the middleware ?

Not sure right now how, but I will try to take a look tomorrow with fresh eyes. Anyway, if you have an idea, please feel free to suggest or push :pray:

File: lib/tool_shed/webapp/frontend/src/components/pages/ManageApiKey.vue (hunk 7 above)
Comment (davelopez) [reply lib/tool_shed/webapp/frontend/src/components/pages/ManageApiKey.vue]: I've done a second round of refactoring to the Tool Shed API calls in an attempt to handle the errors consistently. I hope this is in better shape.

**Review by mvdbeek — APPROVED** [review]
Let's do this. Thanks a lot for working through all of the details, this is great!

**Comment by mvdbeek** [comment]
I'll leave this open till tomorrow to give others a chance to review

**Review by dannon — APPROVED** [review]
This looks great to me, thanks so much @davelopez for picking this up and running with it.
